CI test was performed after the rebase(https://github.com/bpfkorea/agora/pull/1046) but failed to pass ubuntu-18.04 ldc-master.
Previously, errors only occurred in macOS-10.15,ldc-1.22.0 

https://github.com/bpfkorea/agora/pull/1038/checks?check_run_id=896701409

````
+ dchatty=1
+ timeout -s SEGV 8m ./build/agora-unittests
./build/agora-unittests(+0x772b47)[0x5608d25c2b47]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x128a0)[0x7f6f86a6e8a0]
/lib/x86_64-linux-gnu/libpthread.so.0(pthread_cond_wait+0x243)[0x7f6f86a699f3]
./build/agora-unittests(+0x772f20)[0x5608d25c2f20]
./build/agora-unittests(+0x7602ba)[0x5608d25b02ba]
./build/agora-unittests(+0x760028)[0x5608d25b0028]
./build/agora-unittests(+0x1b2ee3)[0x5608d2002ee3]
./build/agora-unittests(+0x1b1960)[0x5608d2001960]
./build/agora-unittests(+0x7729d1)[0x5608d25c29d1]
./build/agora-unittests(+0x77e4cb)[0x5608d25ce4cb]
./build/agora-unittests(+0x77e3f3)[0x5608d25ce3f3]
./build/agora-unittests(+0x77e24e)[0x5608d25ce24e]
./build/agora-unittests(+0x1e6865)[0x5608d2036865]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f6f860d6b97]
./build/agora-unittests(+0x7c65a)[0x5608d1ecc65a]
##[error]The operation was canceled.
```

--- a/source/agora/api/handler/BlockExternalizedHandler.d
+++ b/source/agora/api/handler/BlockExternalizedHandler.d
@@ -1,0 +1,38 @@
+/*******************************************************************************
+
+    Definitions of the BlockExternalizedHandler
+
+    Copyright:
+        Copyright (c) 2019-2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.api.handler.BlockExternalizedHandler;
+
+import agora.consensus.data.Block;
+
+import vibe.web.rest;
+import vibe.http.common;
+
+public interface BlockExternalizedHandler
+{
+// The REST generator requires @safe methods
+@safe:
+
+    /***************************************************************************
+
+        Externalize block data in JSON format with HTTP POST
+
+        API:
+            See config.event_handlers.block_externalized_handler_addresses
+
+    ***************************************************************************/
+
+    @method(HTTPMethod.POST)
+    @path("/")
+    public void pushBlock (const Block block);
+}

--- a/source/agora/api/handler/PreImageReceivedHandler.d
+++ b/source/agora/api/handler/PreImageReceivedHandler.d
@@ -1,0 +1,38 @@
+/*******************************************************************************
+
+    Definitions of the PreImageReceivedHandler
+
+    Copyright:
+        Copyright (c) 2019-2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.api.handler.PreImageReceivedHandler;
+
+import agora.consensus.data.PreImageInfo;
+
+import vibe.web.rest;
+import vibe.http.common;
+
+public interface PreImageReceivedHandler
+{
+// The REST generator requires @safe methods
+@safe:
+
+    /***************************************************************************
+
+        Push a preImage in JSON format with HTTP POST
+
+        API:
+            See config.event_handlers.preimage_updated_handler_addresses
+
+    ***************************************************************************/
+
+    @method(HTTPMethod.POST)
+    @path("/")
+    public void pushPreImage (const PreImageInfo preimage);
+}

--- a/source/agora/network/NetworkManager.d
+++ b/source/agora/network/NetworkManager.d
@@ -23,6 +23,8 @@
 module agora.network.NetworkManager;
 
 import agora.api.Validator;
+import agora.api.handler.BlockExternalizedHandler;
+import agora.api.handler.PreImageReceivedHandler;
 import agora.common.BanManager;
 import agora.consensus.data.Block;
 import agora.consensus.data.Enrollment;
@@ -823,6 +825,62 @@ public class NetworkManager
         // allocates, called infrequently though
         return format("http://%s:%s", this.node_config.address,
             this.node_config.port);
+    }
+
+    /***************************************************************************
+
+        Instantiates a client object implementing `BlockExternalizedHandler`
+
+        In the default implementation, this returns a `BlockExternalizedHandler`
+
+        Params:
+            address = The address (IPv4, IPv6, hostname) of target Server
+        Returns:
+            A BlockExternalizedHandler to communicate with the server
+            at `address`
+
+    ***************************************************************************/
+
+    public BlockExternalizedHandler getBlockExternalizedHandler
+        (Address address)
+    {
+        import vibe.http.client;
+
+        auto settings = new RestInterfaceSettings;
+        settings.baseURL = URL(address);
+        settings.httpClientSettings = new HTTPClientSettings;
+        settings.httpClientSettings.connectTimeout = this.node_config.timeout;
+        settings.httpClientSettings.readTimeout = this.node_config.timeout;
+
+        return new RestInterfaceClient!BlockExternalizedHandler(settings);
+    }
+
+    /***************************************************************************
+
+        Instantiates a client object implementing `PreImageReceivedHandler`
+
+        In the default implementation, this returns a `PreImageReceivedHandler`
+
+        Params:
+            address = The address (IPv4, IPv6, hostname) of target Server
+        Returns:
+            A PreImageReceivedHandler to communicate with the server
+            at `address`
+
+    ***************************************************************************/
+
+    public PreImageReceivedHandler getPreimageReceivedHandler
+        (Address address)
+    {
+        import vibe.http.client;
+
+        auto settings = new RestInterfaceSettings;
+        settings.baseURL = URL(address);
+        settings.httpClientSettings = new HTTPClientSettings;
+        settings.httpClientSettings.connectTimeout = this.node_config.timeout;
+        settings.httpClientSettings.readTimeout = this.node_config.timeout;
+
+        return new RestInterfaceClient!PreImageReceivedHandler(settings);
     }
 }
 

--- a/source/agora/node/FullNode.d
+++ b/source/agora/node/FullNode.d
@@ -467,7 +467,10 @@ public class FullNode : API
         log.trace("Received Preimage: {}", prettify(preimage));
 
         if (this.enroll_man.addPreimage(preimage))
+        {
             this.network.sendPreimage(preimage);
+            this.pushPreImage(preimage);
+        }
     }
 
     /// GET: /preimage

--- a/source/agora/node/FullNode.d
+++ b/source/agora/node/FullNode.d
@@ -154,7 +154,8 @@ public class FullNode : API
         this.enroll_man = this.getEnrollmentManager(config.node.data_dir,
             config.node, params);
         this.ledger = new Ledger(config.node, params, this.utxo_set,
-            this.storage, this.enroll_man, this.pool, onValidatorsChanged);
+            this.storage, this.enroll_man, this.pool, &this.pushBlock,
+            onValidatorsChanged);
         this.exception = new RestException(
             400, Json("The query was incorrect"), string.init, int.init);
 
@@ -169,6 +170,11 @@ public class FullNode : API
             config.event_handlers.preimage_updated_handler_addresses)
             this.preimage_handlers[address] = this.network
                 .getPreimageReceivedHandler(address);
+
+        // Special case
+        // Block externalized handler is set and push for Genesis block.
+        if (this.block_handlers.length > 0 && this.getBlockHeight() == 0)
+            this.pushBlock(GenesisBlock);
     }
 
     /***************************************************************************

--- a/source/agora/node/FullNode.d
+++ b/source/agora/node/FullNode.d
@@ -14,6 +14,8 @@
 module agora.node.FullNode;
 
 import agora.api.FullNode;
+import agora.api.handler.BlockExternalizedHandler;
+import agora.api.handler.PreImageReceivedHandler;
 import agora.consensus.data.Block;
 import agora.common.Amount;
 import agora.common.BanManager;
@@ -106,6 +108,12 @@ public class FullNode : API
     /// If a custom genesis block is set it will be stored here
     private immutable Block genesis_block;
 
+    /// Block Externalized Handler list
+    protected BlockExternalizedHandler[Address] block_handlers;
+
+    /// PreImage Received Handler list
+    protected PreImageReceivedHandler[Address] preimage_handlers;
+
     /***************************************************************************
 
         Constructor
@@ -149,6 +157,18 @@ public class FullNode : API
             this.storage, this.enroll_man, this.pool, onValidatorsChanged);
         this.exception = new RestException(
             400, Json("The query was incorrect"), string.init, int.init);
+
+        // Make `BlockExternalizedHandler` from config
+        foreach (address;
+            config.event_handlers.block_externalized_handler_addresses)
+            this.block_handlers[address] = this.network
+                .getBlockExternalizedHandler(address);
+
+        // Make `PreImageReceivedHandler` from config
+        foreach (address;
+            config.event_handlers.preimage_updated_handler_addresses)
+            this.preimage_handlers[address] = this.network
+                .getPreimageReceivedHandler(address);
     }
 
     /***************************************************************************
@@ -465,5 +485,66 @@ public class FullNode : API
 
     protected void onAcceptedTransaction () @safe
     {
+    }
+
+    /***************************************************************************
+
+        Push the block data to the `block_handlers` target server list
+        set in config.
+
+        Convert block data to JSON serialization and send it POST using Rest.
+
+        Params:
+            block = externalized block data
+
+    ***************************************************************************/
+
+    private void pushBlock (const Block block) @safe
+    {
+        foreach (address, handler; this.block_handlers)
+        {
+            runTask({
+                try
+                {
+                    handler.pushBlock(block);
+                }
+                catch (Exception e)
+                {
+                    log.error("Error sending block height #{} to {} :{}",
+                        block.header.height, address, e);
+                }
+            });
+        }
+    }
+
+    /***************************************************************************
+
+        Push the preimage to the `preimage_handlers` target server list
+        set in config.
+
+        Convert `PreImageInfo` to JSON serialization and send it POST
+        using Rest.
+
+        Params:
+            preImage = Received `PreImageInfo`
+
+    ***************************************************************************/
+
+    private void pushPreImage (const PreImageInfo pre_image) @safe
+    {
+        foreach (address, handler; this.preimage_handlers)
+        {
+            runTask({
+                try
+                {
+                    handler.pushPreImage(pre_image);
+                }
+                catch (Exception e)
+                {
+                    log.error("Error sending preImage (enroll_key: {}) to {} :{}",
+                        pre_image.enroll_key, address, e);
+                }
+            });
+        }
     }
 }

--- a/source/agora/node/Ledger.d
+++ b/source/agora/node/Ledger.d
@@ -76,9 +76,12 @@ public class Ledger
     /// Node config
     private NodeConfig node_config;
 
-    /// If not null call this delegate if the validator set changed after
-    /// a block was externalized
+    /// If not null call this delegate when a new block was added to the ledger
     private void delegate () nothrow @safe onValidatorsChanged;
+
+    /// If not null call this delegate
+    /// A block was externalized
+    private void delegate (const Block) @safe onAcceptedBlock;
 
     /// Parameters for consensus-critical constants
     private immutable(ConsensusParams) params;
@@ -94,6 +97,8 @@ public class Ledger
             storage = the block storage
             enroll_man = the enrollmentManager
             pool = the transaction pool
+            onAcceptedBlock = optional delegate to call
+                              when a block was added to the ledger
             onValidatorsChanged = optional delegate to call after the validator
                                   set changes when a block was externalized
 
@@ -103,6 +108,7 @@ public class Ledger
         NodeConfig node_config, immutable(ConsensusParams) params,
         UTXOSet utxo_set, IBlockStorage storage,
         EnrollmentManager enroll_man, TransactionPool pool,
+        void delegate (const Block) @safe onAcceptedBlock = null,
         void delegate () nothrow @safe onValidatorsChanged = null)
     {
         this.node_config = node_config;
@@ -111,6 +117,7 @@ public class Ledger
         this.storage = storage;
         this.enroll_man = enroll_man;
         this.pool = pool;
+        this.onAcceptedBlock = onAcceptedBlock;
         this.onValidatorsChanged = onValidatorsChanged;
         if (!this.storage.load(GenesisBlock))
             assert(0);
@@ -246,6 +253,7 @@ public class Ledger
         and add all of its outputs to the UTXO set.
         If there are any enrollments in the block,
         add enrollments to the validator set.
+        If not null call the `onAcceptedBlock` delegate.
 
         Params:
             block = the block to add
@@ -275,6 +283,9 @@ public class Ledger
 
         if (this.onValidatorsChanged !is null && validators_changed)
             this.onValidatorsChanged();
+
+        if (this.onAcceptedBlock !is null)
+            this.onAcceptedBlock(block);
     }
 
     /***************************************************************************

--- a/source/agora/test/Base.d
+++ b/source/agora/test/Base.d
@@ -563,6 +563,9 @@ public class TestAPIManager
 
 public class TestNetworkManager : NetworkManager
 {
+    import agora.api.handler.BlockExternalizedHandler;
+    import agora.api.handler.PreImageReceivedHandler;
+
     ///
     public Registry* registry;
 
@@ -606,6 +609,20 @@ public class TestNetworkManager : NetworkManager
         cstring _data_dir)
     {
         return new FakeClockBanManager(conf);
+    }
+
+    ///
+    protected final override BlockExternalizedHandler getBlockExternalizedHandler
+        (Address address)
+    {
+        assert(0, "Not supported");
+    }
+
+    ///
+    protected final override PreImageReceivedHandler getPreimageReceivedHandler
+        (Address address)
+    {
+        assert(0, "Not supported");
     }
 }
 

